src/streaming/zarrv2.array.writer.cpp

```diff
@@ -0,0 +1,177 @@
+#include "macros.hh"
+#include "zarrv2.array.writer.hh"
+#include "sink.creator.hh"
+#include "zarr.common.hh"
+
+#include <nlohmann/json.hpp>
+
+#include <latch>
+#include <stdexcept>
+
+namespace {
+std::string
+sample_type_to_dtype(ZarrDataType t)
+
+{
+    const std::string dtype_prefix =
+      std::endian::native == std::endian::big ? ">" : "<";
+
+    switch (t) {
+        case ZarrDataType_uint8:
+            return dtype_prefix + "u1";
+        case ZarrDataType_uint16:
+            return dtype_prefix + "u2";
+        case ZarrDataType_uint32:
+            return dtype_prefix + "u4";
+        case ZarrDataType_uint64:
+            return dtype_prefix + "u8";
+        case ZarrDataType_int8:
+            return dtype_prefix + "i1";
+        case ZarrDataType_int16:
+            return dtype_prefix + "i2";
+        case ZarrDataType_int32:
+            return dtype_prefix + "i4";
+        case ZarrDataType_int64:
+            return dtype_prefix + "i8";
+        case ZarrDataType_float32:
+            return dtype_prefix + "f4";
+        case ZarrDataType_float64:
+            return dtype_prefix + "f8";
+        default:
+            throw std::runtime_error("Invalid data type: " +
```

Comment:
Are you catching this error somewhere? Why are you using an exception here instead of returning a boolean and logging an error, as you do in the rest of the codebase?

Reply:
Ultimately it bubbles up to `Writer::write_frame()`, and the caller of that would need to catch it. Although I think we can handle it a bit more gracefully.